### PR TITLE
Batch A2C/PPO/DQN

### PR DESCRIPTION
**Pull request by muupan**
Merge ~#328~ ~#339~ ~#340~ ~#341~  first.

- A2C does not support reset with done=False and recurrent models while A3C does
- Batch DQN does not support recurrent models while normal DQN does

Tasks
- [x] clarify the batch interface of agents
- [x] add tests for batch DQN
- [x] check scores of batch PPO
- [x] check scores of batch DQN
- [x] check scores of batch A2C

### CODE REVIEW

**Comment by muupan** [comment]
I confirmed batch training of DQN achieved similar scores to non-batch one.
- Batch DQN x8: `examples/ale/train_dqn_batch_ale.py --num-env 8 --eval-interval 1000000 --env {env_id}`
- DQN: `examples/ale/train_dqn_ale.py --eval-interval 1000000 --env {env_id}`
image
image

**Comment by muupan** [comment]
I confirmed that A2C achieves similar scores to A3C. Since A2C uses a larger batch size than A3C, (5*24 vs 5), the scores can be different.

A2C: examples/ale/train_a2c_ale.py --gpu 0 --num-envs 24 --max-episode-len 27000 --steps 80000000 --eval-interval 1000000 --env {env_id}
A3C: examples/ale/train_a3c_ale.py 24 --max-episode-len 27000 --steps 80000000 --eval-interval 1000000 --env {env_id}

A2C duration: 16:12:13
A3C duration: 25:25:03

image
image

**Comment by muupan** [comment]
I confirmed that all of the below achieve similar scores:
- PPO: new PPO trained via `act_and_train` and `stop_episode_and_train`
- Batch PPO x1: new PPO trained via `batch_act_and_train` and `batch_observe_and_train` with 1 env
- PPO old: PPO before this PR

I also trained new PPO with 8 envs. The scores seems slightly different from those with 1 env, but I think it makes sense since the state distributions in a dataset are different.

Commands:
- Batch PPO x1: `examples/gym/train_ppo_batch_gym.py --num-env 1 --standardize-advantages --eval-interval 10000 --env {env_id}`
- Batch PPO x8: `examples/gym/train_ppo_batch_gym.py --num-env 8 --standardize-advantages --eval-interval 10000 --logger-level 20 --env {env_id}`
- PPO: `examples/gym/train_ppo_gym.py --standardize-advantages --eval-interval 10000 --logger-level 20 --env {env_id}`
- PPO old: `examples/gym/train_ppo_gym.py --normalize-obs --standardize-advantages --eval-interval 10000 --env {env_id}`

Duration on Hopper-v2 (in cpu):
- Batch PPO x1 (1cpu): 01:10:45
- Batch PPO x8 (8cpu): 00:28:45
- PPO (1cpu): 00:44:23
- PPO old (1cpu): 00:42:42

image
image
image

File: chainerrl/agents/a2c.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        with chainer.no_backprop_mode():
+            pout, _ = self.model.pi_and_v(statevar)
+            action = pout.sample().data
+        return chainer.cuda.to_cpu(action)
+
+    def batch_observe_and_train(self, batch_obs, batch_reward, batch_done,
+                                batch_reset):
+
+        if any(batch_reset):
+            warnings.warn('A2C currently does not support resetting an env withtout reaching a terminal state during training. When receiving True in batch_reset, A2C considers it as True in batch_done instead.')  # NOQA
```
[the remark this answers is not in the document]
Comment (muupan) [reply chainerrl/agents/a2c.py]: fixed

File: chainerrl/experiments/train_agent_batch.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+from chainerrl.misc.makedirs import makedirs
+
+
+def train_agent_batch(agent, env, steps, outdir, log_interval=None,
+                      max_episode_len=None, eval_interval=None,
+                      step_offset=0, evaluator=None, successful_score=None,
+                      step_hooks=[], return_window_size=100, logger=None):
+    """Train an agent in a batch environment.
+
+    Args:
```
[the remark this answers is not in the document]
Comment (muupan) [reply chainerrl/experiments/train_agent_batch.py]: fixed

File: chainerrl/agent.py [hunk 3]
```diff
@@ -173,3 +173,68 @@ def process_idx(self):
...
+                current state is not terminal.
+
+
+        Returns:
+            None
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def batch_observe_and_train(self, batch_obs):
```
Comment (toslunar) [inline chainerrl/agent.py]: The args don't match with concrete implementations.

File: chainerrl/agent.py [hunk 4]
```diff
@@ -173,3 +173,68 @@ def process_idx(self):
...
+        Args:
+            batch_obs (Sequence of ~object): Observations.
+
+        Returns:
+            Sequence of ~object: Actions.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def batch_observe(self, batch_obs, batch_reward, batch_done, batch_reset):
```
Comment (toslunar) [inline chainerrl/agent.py]: Being not allowed to learn here, agents don't have to take `batch_reward` and `batch_done` in the `batch_observe` method.  I'm not sure if `batch_obs` can be removed.

File: chainerrl/agent.py (hunk 3 above)
Comment (muupan) [reply chainerrl/agent.py]: fixed.

File: chainerrl/agent.py (hunk 4 above)
Comment (muupan) [reply chainerrl/agent.py]: As we discussed offline, I keep the same arguments to make the API consistent between training and evaluation.

File: chainerrl/env.py [hunk 5]
```diff
@@ -27,3 +27,29 @@ def reset(self):
...
+    def step(self, action):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def reset(self, mask):
+        """Reset envs.
+
+        Args:
+            mask (Sequence of bool): Mask array that specifies which env to
+                skip. If omitted, all the envs are reset.
```
[the remark this answers is not in the document]
Comment (toslunar) [reply chainerrl/env.py]: I'm sorry that I was confused.

File: chainerrl/envs/multiprocess_vector_env.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            remote.send(('seed', seed))
+        results = [remote.recv() for remote in self.remotes]
+        return results
+
+    @property
+    def num_envs(self):
+        return len(self.remotes)
+
+    def _assert_not_closed(self):
+        assert not self.closed, "Trying to operate on a MultiprocessVectorEnv after calling close()"  # NOQA
```
Comment (toslunar) [inline chainerrl/envs/multiprocess_vector_env.py]: The message might be confusing when `close` is called twice.

File: chainerrl/experiments/evaluator.py [hunk 7]
```diff
@@ -62,10 +64,81 @@ def run_evaluation_episodes(env, agent, n_runs, max_episode_len=None,
...
+
+        # Make mask. 0 if done/reset, 1 if pass
+        end = np.logical_or(resets, dones)
+        not_end = np.logical_not(end)
+
+        episode_returns.extend(
+            np.ma.masked_array(episode_r, not_end).compressed())
+        episode_lengths.extend(
+            np.ma.masked_array(episode_len, not_end).compressed())
+        episode_r *= not_end
```
Comment (toslunar) [inline chainerrl/experiments/evaluator.py]: How about `episode_r[end] = 0`?

File: chainerrl/experiments/train_agent_batch.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            #   1. increment the episode count
+            #   2. record the return
+            #   3. clear the record of rewards
+            #   4. clear the record of the number of steps
+            #   5. reset the env to start a new episode
+            episode_idx += end
+            recent_returns.extend(
+                np.ma.masked_array(episode_r, not_end).compressed())
+            episode_r *= not_end
+            episode_len *= not_end
```
Comment (toslunar) [inline chainerrl/experiments/train_agent_batch.py]: ditto

File: chainerrl/experiments/evaluator.py [hunk 9]
```diff
@@ -62,10 +64,81 @@ def run_evaluation_episodes(env, agent, n_runs, max_episode_len=None,
...
+            resets = (episode_len == max_episode_len)
+        # Agent observes the consequences
+        agent.batch_observe(obss, rs, dones, resets)
+
+        # Make mask. 0 if done/reset, 1 if pass
+        end = np.logical_or(resets, dones)
+        not_end = np.logical_not(end)
+
+        episode_returns.extend(
+            np.ma.masked_array(episode_r, not_end).compressed())
```
Comment (toslunar) [inline chainerrl/experiments/evaluator.py]: How about `.extend(episode_r[end])`?

File: chainerrl/replay_buffer.py [hunk 10]
```diff
@@ -402,20 +402,21 @@ def stop_current_episode(self):
...
             [elem['next_state'] for elem in experiences], xp, phi),
-        'next_action': xp.asarray(
-            [elem['next_action'] for elem in experiences]),
         'is_state_terminal': xp.asarray(
             [elem['is_state_terminal'] for elem in experiences],
-
             dtype=np.float32)}
+    if all(elem['next_action'] is not None for elem in experiences):
+        batch_exp['next_action'] = xp.asarray(
+            [elem['next_action'] for elem in experiences])
```
Comment (toslunar) [inline chainerrl/replay_buffer.py]: Should we check all `next_action`s are `None` otherwise?

Comment (muupan) [reply chainerrl/replay_buffer.py]: If some of `elem['next_action']` is None, then `cupy.asarray` fails due to `Value Error: Unsupported dtype object`. So I batch them only if all of them are not None.

File: chainerrl/envs/multiprocess_vector_env.py (hunk 6 above)
Comment (muupan) [reply chainerrl/envs/multiprocess_vector_env.py]: simplified the message.

File: chainerrl/experiments/evaluator.py (hunk 7 above)
Comment (muupan) [reply chainerrl/experiments/evaluator.py]: 👍

File: chainerrl/experiments/train_agent_batch.py (hunk 8 above)
Comment (muupan) [reply chainerrl/experiments/train_agent_batch.py]: 👍

File: chainerrl/experiments/evaluator.py (hunk 9 above)
Comment (muupan) [reply chainerrl/experiments/evaluator.py]: 👍

File: chainerrl/replay_buffer.py (hunk 10 above)
Comment (toslunar) [reply chainerrl/replay_buffer.py]: OK.  I understand the design that a replay buffer makes `elem['next_action']` available if it'll be available as `xp.ndarray`.

**Pull request merged**